[FIX] IAM with dynamic trigger showing forever

jinliu9508 opened the pull request:
# Description
## One Line Summary
Fix an issue with IAM with dynamic triggers like "SESSION_TIME" showing up repetitively after being dismissed.

## Details

### Motivation
Fix the incorrect app behavior when IAM with both custom trigger and session_time trigger continues to show up even it was dismissed.

### Scope
- This issue happens strictly when both a time based trigger like Session_Time and an in-app trigger of "IS NOT EQUAL" exist with AND. The first time both triggers evaluated to true, IAM shows up, gets dismissed, and will reevaluate the triggers again. In the reevaluation, the system first determines if there is only dynamic trigger. If there is, it forces evaluate to false to suppress the IAM from showing up again. But in this case, another in-app trigger exists and will always evaluate to true due to the IS NOT EQUAL characteristics we currently have, thus a repeating IAM occurs.
- This PR alters the logic on when to redisplay message so that we no longer make redisplay messages available to avoid duplicate redisplay of a same message. A message can be redisplayed if at least one trigger has changed or a new trigger is added when a dynamic trigger exists.

## Manual testing
Before the fix:
IAM setting to reproduce the issue:
![image](https://github.com/OneSignal/OneSignal-Android-SDK/assets/14221056/4c5c4272-3d8a-4bd8-a09d-81aa593d4b1d)
This IAM would show up 5 seconds after app start up and continue to show up after being dismissed.

After the fix:
1. IAM with whatever IS NOT 5 AND Session_Time > 5 with Every time trigger conditions are satisfied checked:
    - Result: This IAM will show up 5 seconds after app start up and will not show up again once dismissed. Whenever I add a new trigger, the IAM shows up again.
2. IAM with whatever = true AND Session_Time > 5 with Every time trigger conditions are satisfied checked:
    - Result: This IAM will not show up initially, but will show up if I add a trigger whatever = true EVERY TIME after 5 seconds of startup. If I add a trigger with an unrelated key, the IAM will not show.
3. IAM with a single trigger whatever2 = 2
    - Result: This IAM will not show up unless a trigger whatever2 = 2 is added. Any value for the same key that is not equals to 2 will not show the IAM. 
4. IAM with a single dynamic trigger Session_Time > 5
    - Result: This IAM will only show once 5 seconds after startup.

# Affected code checklist
   - [ ] Notifications
      - [ ] Display
      - [ ] Open
      - [ ] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [x] In-App Messaging
   - [ ] REST API requests
   - [ ] Public API changes

# Checklist
## Overview
   - [x] I have filled out all **REQUIRED** sections above
   - [x] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [x] I have included test coverage for these changes, or explained why they are not needed
   - [ ] All automated tests pass, or I explained why that is not possible
   - [x] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [x] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [x] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/OneSignal/OneSignal-Android-SDK/2137)
<!-- Reviewable:end -->

jkasten2 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/InAppMessagesManager.kt:
```diff
@@ -450,13 +450,14 @@ internal class InAppMessagesManager(
             if (!message.isTriggerChanged && redisplayedInAppMessages.contains(message) &&
-                _triggerController.isTriggerOnMessage(message, newTriggersKeys)
+                _triggerController.isTriggerOnMessage(message, newTriggersKeys) &&
+                !message.isDisplayedInSession
```
This breaks the "Every time trigger conditions are satisfied" logic so we shouldn't do this.

jinliu9508 replied:
Ok, I just pushed a different change that only target dynamic trigger firing logic.

jkasten2 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/triggers/impl/DynamicTriggerController.kt:
```diff
@@ -44,14 +44,31 @@ internal class DynamicTriggerController(
+                        } else {
+                            _time.currentTimeMillis - lastTimeAppDismissed
+                        }
+                    if (evaluateTimeIntervalWithOperator(
+                            lastTimeDismissedInterval.toDouble(),
+                            currentTimeInterval.toDouble(),
+                            trigger.operatorType,
+                        )
+                    ) {
+                        return false
+                    }
```
Won't this break the "Every time trigger conditions are satisfied" if the IAM had a custom trigger with AND?

jinliu9508 replied:
Yes it does. I pushed a new approach with memorization.

jkasten2 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/triggers/impl/DynamicTriggerController.kt:
```diff
@@ -44,14 +44,31 @@ internal class DynamicTriggerController(
             }
 
             var currentTimeInterval: Long = 0
+            val lastTimeAppDismissed = _state.lastTimeInAppDismissed
```
`lastTimeAppDismissed` is missing `In` in the name, should be `lastTimeInAppDismissed`

jinliu9508 replied:
Outdated due to rework

jkasten2 reviewed (changes requested):
This latest solution still looks like it would break the "Every time trigger conditions are satisfied" behavior. We want to keep the behavior where the developer should be able to reshow the IAM by calling `addTrigger` again. Can you test this scenario and add what you tested to the PR notes?

jinliu9508 commented:
> This latest solution still looks like it would also break the "Every time trigger conditions are satisfied" behavior. We want to keep the behavior where the developer should be able to reshow the IAM by calling `addTrigger` again with the same key. Can you test this scenario and add what you tested to the PR notes?

The latest change adds a set that stores evaluated combinations of trigger to prevent a certain combination of trigger from showing multiple time. The set will get cleared every time addTrigger or removeTrigger is called. And it appears in my test that the IAM will show up once per startup and every time I add a new trigger.

jkasten2 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/triggers/impl/TriggerController.kt:
```diff
@@ -17,6 +17,7 @@ internal class TriggerController(
     private var _dynamicTriggerController: DynamicTriggerController,
 ) : ITriggerController, IModelStoreChangeHandler<TriggerModel> {
     val triggers: ConcurrentHashMap<String?, Any?> = ConcurrentHashMap()
+    val evaluatedCombinations = mutableSetOf<String>()
```
This should be at the Message level, not global for all IAMs. Otherwise if new IAMs are fetched when the app is running then they won't evaluate to true when they should.

jkasten2 replied:
I think this is probably fine actually, since trigger.id is a UUID.

jkasten2 reviewed (changes requested):
This new logic has a lot of overlap with the pre-existing `isTriggerChanged` and `isDisplayedInSession` logic. I think we should remove these (if they cover everything) before merging in these changes so we don't have redundant logic that would much harder to reason about later.

jinliu9508 commented:
> This new logic has a lot of overlap with the pre-existing `isTriggerChanged` and `isDisplayedInSession` logic. I think we should remove these (if they cover everything) before merging in these changes so we don't have redundant logic that would much harder to reason about later.

Yes, I also realized that this particular approach uses similar logic as 'isTriggerChanged' so I made a different approach that only alter the 'makeRedisplayMessagesAvailableWithTriggers' logic so that onTriggerCompleted no longer redisplay messages. Instead, it was done by onTriggerConditionChanged before trigger evaluation. I have updated the test part in the description to include some common IAM setup that is potentially affected by the change.

jkasten2 reviewed (changes requested):
>2. IAM with whatever = true AND Session_Time > 5 with Every time trigger conditions are satisfied checked:

There is still a bug with that IAM in the follow scenario.
1. Call `addTriger("whatever", true)`
2. IAM displays as expected
3. Call `addTriger("somethingElse", true)`
4. BUG: IAM displays again when it should not
   - It is expected that it only re-display if `addTriger("whatever", true)` was called again.

jinliu9508 commented:
> > 2. IAM with whatever = true AND Session_Time > 5 with Every time trigger conditions are satisfied checked:
> 
> There is still a bug with that IAM in the follow scenario.
> 
> 1. Call `addTriger("whatever", true)`
> 2. IAM displays as expected
> 3. Call `addTriger("somethingElse", true)`
> 4. BUG: IAM displays again when it should not
>    
>    * It is expected that it only re-display if `addTriger("whatever", true)` was called again.

Nice found! I have fixed this bug and removed some unneeded code for the latest commit.

jkasten2 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/InAppMessagesManager.kt:
```diff
@@ -451,12 +451,19 @@ internal class InAppMessagesManager(
+                redisplayedInAppMessages.contains(message) &&
+                (
+                    _triggerController.isTriggerOnMessage(message, newTriggersKeys) ||
+                        isNewTriggerAdded && _triggerController.messageHasOnlyDynamicTriggers(message)
```
This `if` statement is now well over the line of what is considered readable. Let's split this up so it is more readable before we merge this.

jinliu9508 replied:
Agreed, the if statement is now broken into multiple descriptive boolean values to enhance readability.

jkasten2 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/InAppMessagesManager.kt:
```diff
@@ -451,13 +451,17 @@ internal class InAppMessagesManager(
      * Make all messages with redisplay available if:
      * - Already displayed
-     * - At least one Trigger has changed
+     * - At least one existing Trigger has changed OR a new trigger is added when there is only dynamic trigger
```
👍 on updating the comments!

jkasten2 reviewed (approved):
No action needed in this PR, but noted here for the future. I think there is still more states the can be cleaned up to simplify this manager's logic. Some that come to mind are:
1. I don't think `messageHasOnlyDynamicTriggers` needs to be a concept
2. `redisplayedInAppMessages` overlaps with `message.isDisplayedInSession`

jinliu9508 commented on OneSignalSDK/onesignal/in-app-messages/src/main/java/com/onesignal/inAppMessages/internal/InAppMessagesManager.kt:
```diff
@@ -451,13 +451,17 @@ internal class InAppMessagesManager(
-            if (!message.isTriggerChanged && redisplayedInAppMessages.contains(message) &&
-                _triggerController.isTriggerOnMessage(message, newTriggersKeys)
-            ) {
+            val isMessageDisplayed = redisplayedInAppMessages.contains(message)
```
Ok I just fixed this.
